### PR TITLE
Classify TiFlash and TiKV in PeerStores

### PR DESCRIPTION
**Pull request by handlerww**
### What problem does this PR solve? 
Classify TiFlash and TiKV in peerStores to avoid misleading information in tc.Status.PD.PeerStores and tc.Status.TiFlash.PeerStores
### What is changed and how does it work?
TiKV member manager and TiFlash member manager use regex to select incluster stores. However, they put mismatch results all in peerStores. It needs to filter out the components not belongs to them.
### Check List 

Tests 

 - Unit test
 - Manual test
```yaml
apiVersion: pingcap.com/v1alpha1
kind: TidbCluster
metadata:
  name: basic
  namespace: pingcap
spec:
  version: v4.0.8
  timezone: UTC
  pvReclaimPolicy: Delete
  enableDynamicConfiguration: true
  configUpdateStrategy: RollingUpdate
  clusterDomain: "cluster.local"
  discovery: {}
  pd:
    baseImage: registry.cn-beijing.aliyuncs.com/tidb/pd
    replicas: 1
    requests:
      storage: "1Gi"
    config: 
      schedule:
        max-store-down-time: 1m
  tikv:
    baseImage: registry.cn-beijing.aliyuncs.com/tidb/tikv
    replicas: 3
    requests:
      storage: "1Gi"
    config: {}
  tidb:
    baseImage: registry.cn-beijing.aliyuncs.com/tidb/tidb
    replicas: 1
    service:
      type: ClusterIP
    config: {}
  tiflash:
    baseImage: pingcap/tiflash
    maxFailoverCount: 3
    replicas: 1
    storageClaims:
    - resources:
        requests:
          storage: 1Gi
    config: {}
```
```yaml
apiVersion: pingcap.com/v1alpha1
kind: TidbCluster
metadata:
  name: cluster2
  namespace: pingcap
spec:
  version: v4.0.8
  timezone: UTC
  pvReclaimPolicy: Delete
  enableDynamicConfiguration: true
  configUpdateStrategy: RollingUpdate
  clusterDomain: "cluster.local"
  cluster:
    name: basic
    namespace: pingcap
    clusterDomain: "cluster.local"
  discovery: {}
  pd:
    baseImage: registry.cn-beijing.aliyuncs.com/tidb/pd
    replicas: 2
    requests:
      storage: "1Gi"
    config: 
      schedule:
        max-store-down-time: 1m
  tikv:
    baseImage: registry.cn-beijing.aliyuncs.com/tidb/tikv
    replicas: 1
    requests:
      storage: "1Gi"
    config: {}
  tidb:
    baseImage: registry.cn-beijing.aliyuncs.com/tidb/tidb
    replicas: 0
    service:
      type: ClusterIP
    config: {}
```
Create two clusters according to these 2 yamls.

```shell
kubectl describe tc basic
```
You can see that the TiKV and TiFlash PeerStores are classified in right place.
image
image

Code changes

 - Has Go code change

Side effects

 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

File: pkg/manager/member/tiflash_member_manager.go [hunk 1]
```diff
@@ -661,7 +661,11 @@ func (m *tiflashMemberManager) syncTidbClusterStatus(tc *v1alpha1.TidbCluster, s
 		if store.Store != nil && pattern.Match([]byte(store.Store.Address)) {
 			stores[status.ID] = *status
 		} else {
```
Comment (DanielZhangQD) [inline pkg/manager/member/tiflash_member_manager.go:663]: `store.Store` can be `nil` in this case.

File: pkg/manager/member/tiflash_member_manager.go [hunk 2]
```diff
@@ -661,7 +661,11 @@ func (m *tiflashMemberManager) syncTidbClusterStatus(tc *v1alpha1.TidbCluster, s
...
-			peerStores[status.ID] = *status
+			storePodName := strings.Split(store.Store.Address, ".")[0]
+			storeKind := strings.Split(storePodName, "-")
+			if storeKind[len(storeKind)-2] == label.TiFlashLabelVal {
```
Comment (DanielZhangQD) [inline pkg/manager/member/tiflash_member_manager.go:666]: Define a function `matchStore(store.Store.Address, label.TiFlashLabelVal)`, which can also be used in #3514, #3513.

File: pkg/manager/member/tikv_member_manager.go [hunk 3]
```diff
@@ -680,7 +680,11 @@ func (m *tikvMemberManager) syncTidbClusterStatus(tc *v1alpha1.TidbCluster, set
 		if store.Store != nil && pattern.Match([]byte(store.Store.Address)) {
 			stores[status.ID] = *status
 		} else {
-			peerStores[status.ID] = *status
+			storePodName := strings.Split(store.Store.Address, ".")[0]
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_member_manager.go:683]: Same comments as those for tiflash.

File: pkg/manager/member/tiflash_member_manager.go (hunk 2 above)
Comment (DanielZhangQD) [reply pkg/manager/member/tiflash_member_manager.go:666]: Please also update UT.

File: pkg/manager/member/tiflash_member_manager.go [hunk 4]
```diff
@@ -659,10 +659,12 @@ func (m *tiflashMemberManager) syncTidbClusterStatus(tc *v1alpha1.TidbCluster, s
 			status.LastTransitionTime = oldStore.LastTransitionTime
 		}
 
-		if store.Store != nil && pattern.Match([]byte(store.Store.Address)) {
-			stores[status.ID] = *status
-		} else {
-			peerStores[status.ID] = *status
+		if store.Store != nil {
```
Comment (DanielZhangQD) [inline pkg/manager/member/tiflash_member_manager.go:662]: Please update UT to cover the code change.

File: pkg/pdapi/pd_control.go [hunk 5]
```diff
@@ -180,3 +180,7 @@ func NewFakePDControl(kubeCli kubernetes.Interface) *FakePDControl {
 func (fpc *FakePDControl) SetPDClient(namespace Namespace, tcName string, pdclient PDClient) {
 	fpc.defaultPDControl.pdClients[pdClientKey("http", namespace, tcName)] = pdclient
 }
+
+func (fpc *FakePDControl) SetPDClientForCrossRegion(namespace Namespace, tcName string, tcClusterDomain string, pdclient PDClient) {
```
Comment (DanielZhangQD) [inline pkg/pdapi/pd_control.go:184]:
```suggestion
func (fpc *FakePDControl) SetPDClientWithClusterDomain(namespace Namespace, tcName string, tcClusterDomain string, pdclient PDClient) {
```

File: pkg/manager/member/tikv_member_manager_test.go [hunk 6]
```diff
@@ -1482,6 +1482,72 @@ func TestTiKVMemberManagerSyncTidbClusterStatus(t *testing.T) {
 				g.Expect(tc.Status.TiKV.Synced).To(BeTrue())
 			},
 		},
+		{
+			name: "get tiflash Store and PeerStores without TiFlash Stores",
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_member_manager_test.go:1486]:
```suggestion
			name: "get TiKV Store and PeerStores without TiFlash Stores",
```

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Comment by dragonly** [comment]
maybe the manual test can be more detailed, so that anyone out there can validate by reproduce your test procedures.

rest LGTM.

**Comment by dragonly** [comment]
LGTM

**Review by dragonly — APPROVED** [review]

**Pull request merged**